### ISSUE TITLE
Make Python 2-3 compatible

### ISSUE THREAD

**Issue by markgw** [markgw/pimlico#8]
We really need Pimlico to be compatible with both Python 2 and Python 3. Otherwise, people who use Python 3 will be put off from using Pimlico because they'd have to write their own modules in Python 2, or even worse if they're using their previously existing Python 3 code within Pimlico modules. For the same reasons, compatibility is better than just moving to Python 3.

This is a major change that will affect every Python source file in the project. It should be carried out in a branch and probably only after I've implemented much more comprehensive testing (#7), so I can find where it breaks things.

I think that most of the libraries Pimlico uses are 2-3 compatible, but should check this: if there are much-used libraries that don't support Python 3, we'll end up with modules or features that only work if you're using Python 2, which would be nasty.

**Comment by markgw** [comment]
I plan to use the [future](http://python-future.org/index.html) library for this.

We will use the 2-stage conversion process. Run futurize --stage1 first on the entire codebase. Then run stage2 and check all of the changes.

Because of the complexity of the codebase and some of the fancy importing stuff we do, it's possible stage2 could break things, so it will be important to ensure that those changes are tested in Py2 and 3.

This change is becoming a high priority, as it's stopping people from using Pimlico.

**Comment by markgw** [comment]
To be addressed in [python3 branch](/markgw/pimlico/tree/python3).

Module tests and other unit tests are far from comprehensive, but there's probably enough to get started on this and I need to get started asap, as it's blocking people from using Pimlico.

**Comment by markgw** [comment]
Steps carried out to get started:
- Install `future` in the default virtualenv
- Create another virtualenv using python 3, so we can test converted code:
  `virtualenv -p python3 lib/virtualenv/py3`
  This can now be used for Pimlico by adding the `PIMENV` env var:
  `PIMENV=py3 ./pimlico.sh ...`
- Try using Py3 virtualenv: `PIMENV=py3 ./all_test_pipelines.sh`
  This fails, as our code is not yet syntactically valid Python 3, of course.
- Activate default virtualenv to make futurize available:
  `. ./lib/virtualenv/default/bin/activate`
- Run stage 1 conversion of all code using `futurize`, to make it more Python 3-like:
  `futurize --stage1 -w -n src/python/pimlico/`
- This seemed to work fine. Run all test pipelines:
  `cd bin`
  `./all_test_pipelines.sh`
- There were a few more things that needed to be resolved to get all the tests passing. In particular, I added some new features to Python dependency checking and installation that allowed specification of versions for some of the pre-defined dependencies.

The entire codebase is now Python 3-like, but not yet runnable in Python 3.

Next we'll need to run [stage 2](http://python-future.org/automatic_conversion.html#stage-2-py3-style-code-with-wrappers-for-py2), which is more dangerous and will require a lot of work to make sure everything is working.

**Comment by markgw** [comment]
Work now continues on stage 2. 

The command `futurize --stage2 -w src/python` applies potentially dangerous conversions to make the code Python 3 compatible. I'll be reviewing the changes it's made manually and committing them as I go.

After that, I will run pipeline and unit tests and get things working. This means that, until then, the code might not be usable.

**Comment by markgw** [comment]
Working through all the files now. To-do list:
- [ ] core/dependencies/base.py
- [ ] core/dependencies/java.py 
- [ ] core/dependencies/python.py
- [ ] core/dependencies/versions.py 
- [ ] core/config.py 
- [ ] cli/browser/tools/corpus.py
- [ ] cli/browser/tools/files.py
- [ ] cli/browser/tools/formatter.py
- [ ] cli/clean.py
- [ ] cli/debug/stepper.py
- [ ] cli/loaddump.py
- [ ] cli/locations.py
- [ ] cli/newmodule.py
- [ ] cli/pyshell.py
- [ ] cli/recover.py
- [ ] cli/reset.py
- [ ] cli/run.py
- [ ] cli/shell/base.py
- [ ] cli/shell/commands.py
- [ ] cli/status.py
- [ ] cli/subcommands.py
- [ ] cli/util.py
- [ ] core/external/java.py
- [ ] core/modules/base.py
- [ ] core/modules/execute.py
- [ ] core/modules/inputs.py
- [ ] core/modules/map/__init__.py
- [ ] core/modules/map/filter.py
- [ ] core/modules/map/multiproc.py
- [ ] core/modules/map/threaded.py
- [ ] core/modules/multistage.py
- [ ] core/modules/options.py
- [ ] datatypes/arrays.py
- [ ] datatypes/base.py
- [ ] datatypes/core.py
- [ ] datatypes/corpora/base.py
- [ ] datatypes/corpora/data_points.py
- [ ] datatypes/corpora/floats.py
- [ ] datatypes/corpora/grouped.py
- [ ] datatypes/corpora/ints.py
- [ ] datatypes/corpora/json.py
- [ ] datatypes/corpora/table.py
- [ ] datatypes/corpora/tokenized.py
- [ ] datatypes/dictionary.py
- [ ] datatypes/embeddings.py
- [ ] datatypes/features.py
- [ ] datatypes/files.py
- [ ] datatypes/gensim.py
- [ ] datatypes/sklearn.py
- [ ] modules/candc/execute.py
- [ ] modules/corenlp/wrapper.py
- [ ] modules/corpora/concat/info.py
- [ ] modules/corpora/group/info.py
- [ ] modules/corpora/interleave/info.py
- [ ] modules/corpora/shuffle/execute.py
- [ ] modules/corpora/subset/info.py
- [ ] modules/corpora/vocab_counter/execute.py
- [ ] modules/embeddings/dependencies/execute.py
- [ ] modules/embeddings/store_word2vec/execute.py
- [ ] modules/embeddings/word2vec/execute.py
- [ ] modules/features/term_feature_compiler/execute.py
- [ ] modules/features/term_feature_matrix_builder/execute.py
- [ ] modules/features/vocab_builder/execute.py
- [ ] modules/gensim/lda_doc_topics/execute.py
- [ ] modules/gensim/utils.py
- [ ] modules/input/embeddings/fasttext/execute.py
- [ ] modules/input/embeddings/glove/read.py
- [ ] modules/input/embeddings/word2vec/execute.py
- [ ] modules/input/text/raw_text_archives/info.py
- [ ] modules/input/text/raw_text_files/info.py
- [ ] modules/input/text_annotations/vrt/info.py
- [ ] modules/opennlp/coreference/execute.py
- [ ] modules/opennlp/coreference/info.py
- [ ] modules/opennlp/coreference_pipeline/execute.py
- [ ] modules/opennlp/coreference_pipeline/info.py
- [ ] modules/opennlp/ner/execute.py
- [ ] modules/opennlp/ner/ner.py
- [ ] modules/opennlp/pos/execute.py
- [ ] modules/opennlp/pos/tagger.py
- [ ] modules/opennlp/tokenize/tokenizer.py
- [ ] modules/output/text_corpus/execute.py
- [ ] modules/regex/annotated_text/execute.py
- [ ] modules/sklearn/logistic_regression/execute.py
- [ ] modules/sklearn/matrix_factorization/execute.py
- [ ] modules/sklearn/matrix_factorization/info.py
- [ ] modules/text/simple_tokenize/info.py
- [ ] modules/text/untokenize/info.py
- [ ] modules/utility/collect_files/execute.py
- [ ] modules/utility/collect_files/info.py
- [ ] modules/visualization/bar_chart/execute.py
- [ ] modules/visualization/bar_chart/plot_template.py
- [ ] modules/visualization/embeddings_plot/execute.py
- [ ] modules/visualization/embeddings_plot/plot_template.py
- [ ] old_datatypes/base.py
- [ ] old_datatypes/caevo.py
- [ ] old_datatypes/core.py
- [ ] old_datatypes/coref/corenlp.py
- [ ] old_datatypes/coref/opennlp.py
- [ ] old_datatypes/dictionary.py
- [ ] old_datatypes/documents.py
- [ ] old_datatypes/embeddings.py
- [ ] old_datatypes/features.py
- [ ] old_datatypes/files.py
- [ ] old_datatypes/floats.py
- [ ] old_datatypes/formatters/tokenized.py
- [ ] old_datatypes/ints.py
- [ ] old_datatypes/keras.py
- [ ] old_datatypes/parse/__init__.py
- [ ] old_datatypes/parse/candc.py
- [ ] old_datatypes/parse/dependency.py
- [ ] old_datatypes/plotting.py
- [ ] old_datatypes/r.py
- [ ] old_datatypes/table.py
- [ ] old_datatypes/tar.py
- [ ] old_datatypes/vrt.py
- [ ] old_datatypes/word_annotations.py
- [ ] test/pipeline.py
- [ ] utils/communicate.py
- [ ] utils/core.py
- [ ] utils/docs/commandgen.py
- [ ] utils/docs/modulegen.py
- [ ] utils/docs/rest.py
- [ ] utils/email.py
- [ ] utils/filesystem.py
- [ ] utils/format.py
- [ ] utils/network.py
- [ ] utils/pipes.py
- [ ] utils/pos.py
- [ ] utils/probability.py
- [ ] utils/progress.py
- [ ] utils/strings.py
- [ ] utils/web.py

**Comment by markgw** [comment]
About halway through the manual checks on stage 2 conversion at 4415c00e70448b9d3b83c92095a4f72bb5664e78.

Now got a lot of pipeline tests running. Some are still failing, of course:
- pipelines/corpora/shuffle.conf[shuffle]
- pipelines/input/fasttext.conf[vectors]
- pipelines/input/glove.conf[vectors]
- pipelines/embeddings/word2vec.conf[word2vec]
- pipelines/embeddings/store_word2vec.conf[store]
- pipelines/embeddings/store_tsv.conf[store]

**Comment by markgw** [comment]
Finished all manual checks of automatic stage 2 conversion in e0b888e.

Still the same pipeline tests failing given above.

Next steps:
- Get all pipeline tests working in Py2
- Get all unit tests working in Py2
- Try running in Py3 and start addressing problems there

**Comment by markgw** [comment]
863c43d does stage 1 futurize for unit tests.

44588b4 does stage 2 futurize.

**Comment by markgw** [comment]
All unit tests and test pipelines passing in Python 2, from d60fc0cbf457706826f94a01bc34360b36764c27.

Next step is to work out how to specify using Python 3 or 2 for virtualenv creation process.

Then allow tests (unit and pipeline) to be run for both Py 2 and 3.

Then run the tests in Py 3 and address issues.

**Comment by markgw** [comment]
All test pipelines passing in both **Python 2 and 3**, as of 8d7047f4e31c92fa47b158dd82ea6c640bb26a88!

Unit tests not working yet. (Mostly for Python 2, but not at all for Python 3.)

**Comment by markgw** [comment]
All tests now passing on Py 2 and 3!

**Comment by markgw** [comment]
There are still some modules that need to be updated to the new datatypes system and these also need to be tested in Python 2&3. However, other than these (which need to be fixed anyway), everything seems to be working and pretty stable now.

Merge python3 branch in afd2ff5db87321a96d4dee63b5da558dfc6eb7ee.

Closing this issue, since this project is pretty much done! Woopee!
